starlette/_utils.py

```diff
@@ -0,0 +1,12 @@
+import asyncio
+import functools
+import typing
+
+
+def iscoroutinefunction(obj: typing.Any) -> bool:
+    while isinstance(obj, functools.partial):
+        obj = obj.func
+
+    return asyncio.iscoroutinefunction(obj) or (
```

Comment:
Hi @Kludex, would it make sense to use [`inspect.iscoroutinefunction`](https://docs.python.org/3/library/inspect.html?highlight=inspect#inspect.iscoroutinefunction) here?

I understand why we cannot just use the said function alone, since `async def __call__(...)` is not captured by it:

```
>>> import inspect
>>> class A:
...     async def __call__(self):
...             ...
... 
>>> inspect.iscoroutinefunction(A())
False
```

I have looked at the code  of [`asyncio.iscoroutinefunction`](https://github.com/python/cpython/blob/3.10/Lib/asyncio/coroutines.py#L164) and it only additionally check for the deprecated `@coroutine` decorator.

Cheers,
Libor

Reply:
You might want to handle functools.wraps as well (you can check obj.__wrapped__)